source/simulator/helper_functions.cc

```diff
@@ -2403,6 +2404,260 @@ namespace aspect
+
+          for (unsigned int q=0; q<n_q_points; ++q)
+            {
+              // Get parameters for the density kernel
+              const Tensor<1,dim> velocity_adjoint = in_adjoint.velocity[q];
+              const Tensor<1,dim> gravity = gravity_model->gravity_vector(in.position[q]);
+
+              // Get parameters for the viscosity kernel
+              const double eta = out.viscosities[q];
+              const SymmetricTensor<2,dim> strain_rate_forward = in.strain_rate[q] - 1./3 * trace(in.strain_rate[q]) * unit_symmetric_tensor<dim>();
+              const SymmetricTensor<2,dim> strain_rate_adjoint = in_adjoint.strain_rate[q] - 1./3 * trace(in_adjoint.strain_rate[q]) * unit_symmetric_tensor<dim>();
+
+              for (unsigned int k=0; k<dofs_per_cell; ++k)
+                {
+                  phi_rho[k] = fe_values[introspection.extractors.compositional_fields[density_idx]].value(k,q);
+                  phi_eta[k] = fe_values[introspection.extractors.compositional_fields[viscosity_idx]].value(k,q);
+                  // we will need the values of the 0th compositional field for the mass matrix assembly below.
+                  phi_0[k] = fe_values[introspection.extractors.compositional_fields[0]].value(k,q);
+                }
+
+              const double JxW = fe_values.JxW(q);
+
+              for (unsigned int i = 0; i<dofs_per_cell; ++i)
+                {
+                  // Calculate the density kernel
+                  local_rhs(i) += -(gravity * velocity_adjoint) * phi_rho[i] * JxW;
+
+                  // Calculate the viscosity kernel
+                  local_rhs(i) += (2 * eta * strain_rate_forward * strain_rate_adjoint) * phi_eta[i] * JxW;
+                  // it's fine to add both of them because only one of them at a time is nonzero
+
+                  for (unsigned int j = 0; j<dofs_per_cell; j++)
+                    {
+                      // Assemble Mass Matrix for the first compositional field. the mass
+                      // matrix for the second compositional field would look exactly the
+                      // same, so we can avoid building it for efficiency. (in fact we
+                      // also would have to deal with the fact that no sparsity pattern
+                      // and memory is allocated for the second compositional field; but
+                      // because we don't build the matrix, we don't care here.)
+                      local_mass_matrix(i,j) += phi_0[i] * phi_0[j] * JxW;
+                    }
+                }
+            }
+
+
+          // Assemble the surface term
+
+          // Note that the solution at the surface does currently not agree with the benchmark.
+          // This could be because of an error here or in the assembly of the RHS in assemblers/adjoint.cc
+          // Given that the volume kernel already spikes at the surface, the issue is likely not (just) the surface
+          // term that is assembled here.
```

Comment:
Haha I know, frowny face indeed. Currently, the issue definitely originates in the RHS assembly, because even if I don't add the surface term here, I get adjoint strainrates that are too high in the surface cell and hence a viscosity and density update that is too large in the surface cell. Of course, that doesn't exclude that the issue is here, because maybe the surface term - if calculated correctly - corrects for this. The surface term is currently very small and doesn't really affect the viscosity / density update.